### PR TITLE
Implement a funsor-based poutine.collapse

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2618 
~~Blocked by https://github.com/pyro-ppl/funsor/pull/361~~
pair coded with @eb8680

This attempts to implement a `poutine.collapse()` using Funsors, in a way that is compatible with Pyro 1.0-style inference algorithms. This is a possible intermediate step towards Pyro 2.0-style inference algorithms as in `contrib.funsor` and #2580.

### CODE REVIEW

**Comment by martinjankowiak** [comment]
what are the semantics on this? it tries to integrate everything out in the collapse context?

**Comment by eb8680** [comment]
> what are the semantics on this? it tries to integrate everything out in the collapse context?

It uses Funsor to lazily evaluate and analytically integrate out all latent variables in the context, hides them from any enclosing effect handlers, and introduces a single new `pyro.factor` statement containing the marginal likelihood of all observations in the context.  This final `factor` site is visible to and compatible with existing inference algorithms like HMC or `TraceEnum_ELBO`.

**Comment by fritzo** [comment]
@eb8680 once we get the initial version of this working, let's try to use Funsor adjoints to sample and replay from the collapsed distributions, as @neerajprad implemented in [UncollapseConjugateMessenger](https://github.com/pyro-ppl/pyro/blob/1.4.0/pyro/contrib/conjugate/infer.py#L115) for reading the collapsed values during HMC.

File: pyro/poutine/collapse_messenger.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+# TODO Remove import guard once funsor is a required dependency.
+try:
+    import funsor
+    from funsor.terms import Funsor
+except ImportError:
+    pass
+
+
+class CollapseMessenger(TraceMessenger):
```
Comment (fritzo) [inline pyro/poutine/collapse_messenger.py:18]: add docs

**Comment by fritzo** [comment]
@eb8680 I believe the new tests are still not exercised on travis due to `pytest.imporortskip("funsir")`. I think those should be moved to the `funsor` stage; we'll need to mark the stage of those tests.

File: pyro/poutine/collapse_messenger.py (hunk 1 above)
Comment (eb8680) [reply pyro/poutine/collapse_messenger.py:18]: What happened to the docstring here?

Comment (fritzo) [reply pyro/poutine/collapse_messenger.py:18]: oops, just pushed.

**Comment by fritzo** [comment]
Well it looks like the tests are now [passing on travis](https://api.travis-ci.com/v3/job/386445321/log.txt):
```
...
tests/infer/test_gradient.py::test_collapse_beta_binomial tensor(0.5234) tensor(0.5234)
tensor(0.0356) tensor(0.0356)
PASSED
tests/infer/test_valid_models.py::test_collapse_beta_bernoulli XFAIL
tests/infer/test_valid_models.py::test_collapse_beta_binomial PASSED
tests/infer/test_valid_models.py::test_collapse_beta_binomial_plate XFAIL
```

**Review by eb8680 — APPROVED** [review]

**Pull request merged**